### ISSUE TITLE
ci: collecting system logs also in case of virt-handler failure

### ISSUE THREAD

**Issue by alicefr** [kubevirt/kubevirt#9391]
**Is your feature request related to a problem? Please describe**:
We use virt-handler in order to collect logs like [journal](https://github.com/kubevirt/kubevirt/blob/main/tests/reporter/kubernetes.go#L553-L606) and [audit](https://github.com/kubevirt/kubevirt/blob/main/tests/reporter/kubernetes.go#L341-L399) logs. However, this doesn't work properly if virt-handler has issues and it isn't deployed correctly. There are tests like in the operator suite that might restart virt-handler.  In the case, there are any issues we are completely losing the system logs.
Would it make sense to deploy a separate demonset only for collecting the logs in the CI?

**Describe the solution you'd like**:
Be able to look at system logs also if there are any failure in one of virt-handler pods

**Comment by oshoval** [comment]
Hi, what about using must-gather which deploys a new daemon set indeed?
It can be updated to collect stuff in case some is missing

**Comment by alicefr** [comment]
@oshoval I'm not familiar with must-gather is it already used in the CI?

**Comment by oshoval** [comment]
> @oshoval I'm not familiar with must-gather is it already used in the CI?

It is not used on CI, it is used on openshift, but IIRC can be used on k8s / our CI jobs (which are based kubevirtci) as well

It creates a DS which is priviledged and run scripts that collect lots of info
please try the above command, if you like, on a local cluster

**Comment by alicefr** [comment]
I quickly went through the README, and it doesn't mention syslogs as audit or journal. Those are the information I need. Additionally, it is more the upstream CI, locally the tests are passing

**Comment by oshoval** [comment]
> I quickly went through the README, and it doesn't mention syslogs as audit or journal. Those are the information I need. Additionally, it is more the upstream CI, locally the tests are passing

You can add to the mentioned file whatever you need
and you can just run it as of the reporter on CI as well

**Comment by oshoval** [comment]
See for example this area
https://github.com/kubevirt/must-gather/blob/main/collection-scripts/gather_nodes#L112

**Comment by alicefr** [comment]
> See please for example this area https://github.com/kubevirt/must-gather/blob/main/collection-scripts/gather_nodes#L112
> 
> I believe you can add there whatever you need Deploy the DS, let it sleep so it wont clean it, access the pod, and see what commands you can run in order to collect the missing parts
> 
> Alternative is to exec / collect via the node in k8s reporter and not via the virt-handler ?

I'd rather prefer something simpler then the must-gather, if possible. Is the k8s reporter a daemonset available in all the nodes?

**Comment by oshoval** [comment]
> I'd rather prefer something simpler then the must-gather, if possible. Is the k8s reporter a daemonset available in all the nodes?

the k8s reporter is not a DS, it is just set of functions compiled as part of the e2e tests
the easiest way, if it is doable, will to update it to run the commands via the node instead via the virt-handler
after all, the node has access to the journal as well right  ?

**Comment by alicefr** [comment]
> > I'd rather prefer something simpler then the must-gather, if possible. Is the k8s reporter a daemonset available in all the nodes?
> 
> the k8s reporter is not a DS, it is just set of functions compiled as part of the e2e tests the easiest way, if it is doable, will to update it to run the commands via the node instead via the virt-handler after all, the node has access to the journal as well right ?

Sorry for the dummy question, but how do I access the nodes? From what I can see in the tests we only use kubernetes API either to fetch the k8s object or by execing in a pod and run the commands on the node.
Anyway, if we can avoid accessing a pod, even better.

**Comment by alicefr** [comment]
> `./cluster-up/ssh.sh <node_name> [command]`
> 
> It runs a script that ssh the node (VM) via the container grepping `ssh.sh` on kubevirtci can show examples
> 
> You can also have a script on the BM and run this way `.../ssh.sh sudo /bin/bash < /scripts/script_name.sh`

I'm not sure if it is the right command. Ideally, it should be integrated with the [dump command](https://github.com/kubevirt/kubevirt/blob/main/cmd/dump/dump.go). AFAIU, this is what is called when a test failed. Running `ssh` will be very practical but not sure if the functional tests setup can access it.

**Comment by oshoval** [comment]
Right, ssh.go runs `./cluster-up/ssh.sh` as well, so it aint helpful

test.sh is running cluster-up / cluster-sync and then the tests
the tests run https://github.com/kubevirt/kubevirt/blob/main/tests/reporter/kubernetes.go#L109
upon test failure,
there you can run `ssh.sh` manually

one thing that can be nicer, is to copy code from ssh.go into the k8s reporter so it will do it using native go instead of a script, but this is only nice to have.

Note that ssh.sh might act different for different providers (for example with the new k3d provider it need to also include /bin/sh before [command]), but lets reach this and see

**Comment by oshoval** [comment]
We continued on slack.
If you want please sum up the findings / action items that we can do in the future to support it.

**Comment by alicefr** [comment]
@oshoval I still haven't successfully gotten the system logs out of the CI. The CI times out and it doesn't call the function that collects the logs. Probably, I'm missing something in hacking the CI (https://github.com/kubevirt/kubevirt/pull/9177/commits/a11c4f12b0c0a35fb34a43c1c3ed18bdacdd8fe6). This is an [example](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/9177/pull-kubevirt-e2e-k8s-1.24-operator/1634232313065771008) of the CI job.

**Comment by oshoval** [comment]
I think you should add the code to the reporter itself upon test error, in the script it might be too late

**Comment by alicefr** [comment]
@oshoval putting that part in the report is quite hacky and challenging. There are a lot of details that depend on the environment where the tests are deployed. For example, on `kind` we don't use ssh or the container engine used for creating the kubevirtci nodes can vary. Additionally, the name of the nodes depends on the TARGET and not on the K8S_PROVISIONER.
After discussing this with @xpivarc, we probably shouldn't log all the nodes in the cluster for all the failures, as it could be quite considerable large logs.

For my part, I have managed to reproduce the failure locally. However, the fact that we cannot collect systems logs without a functional virt-handler instance is a big limitation.
Would it make sense to loop over the virt-handler pods and create a job for collecting logs only on the nodes where it hasn't been deployed correctly? This could reduce the number of collected logs but at least it improves and covers the issue I'm facing. The drawback with this approach is that we need to deploy a job or pod on a failure and it might take time.

**Comment by oshoval** [comment]
I think it is better to do it via the nodes, the other choices as you said as well are more complex,
We can limit it for vm based providers
btw we do have ssh.sh on kind which act differently

**Comment by rmohr** [comment]
> @oshoval putting that part in the report is quite hacky and challenging. There are a lot of details that depend on the environment where the tests are deployed. For example, on kind we don't use ssh or the container engine used for creating the kubevirtci nodes can vary. Additionally, the name of the nodes depends on the TARGET and not on the K8S_PROVISIONER.
After discussing this with @xpivarc, we probably shouldn't log all the nodes in the cluster for all the failures, as it could be quite considerable large logs.

Yes, limiting log dimensions is imo a big +, so we would log nodes with the following constriants:

 * VMIs are running on a node OR virt-handler is not functioning on  that node (for the second case still a max amount, the chance is high that if 100 nodes have virt-handler crash-looping that they all have the same issue, let's say 5).
 * we continue limiting the log range to the test-time, like we do right now already.

> For my part, I have managed to reproduce the failure locally. However, the fact that we cannot collect systems logs without a functional virt-handler instance is a big limitation.
Would it make sense to loop over the virt-handler pods and create a job for collecting logs only on the nodes where it hasn't been deployed correctly? This could reduce the number of collected logs but at least it improves and covers the issue I'm facing. The drawback with this approach is that we need to deploy a job or pod on a failure and it might take time.

If we deploy them in parallel, it should be relatively fast and we can keep the timeouts relatively low. In general if the cluster itself is healthy they should come up pretty fast.

Sounds overall like a solid plan to me.

**Comment by xpivarc** [comment]
@rmohr  +1. Including the nodes where we see unhealthy `virt-handler`s will be a great improvement.

I would discourage the `ssh` approach as it is less general.
